src/app/modules/spcp/sp.oidc.client.ts

```diff
@@ -0,0 +1,583 @@
+    this.set('expiry', 'expiry', 3300) // set expiry key with TTL of 55 minutes, to trigger refresh up to 5min ahead (note that expiry check is done every 60s)
+    return { ndiPublicKeys, baseClient }
+  }
+
+  /**
+   * Method to make network call to retrieve public JWKS from NDI
+   * Max of 3 attemps with timeout of 3s per attempt
+   * @async
+   * @returns NDI's public keys
+   * @throws JwkError if keys are not the correct shape
+   */
+  async retrievePublicKeysFromNdi(): Promise<CryptoKeySet> {
+    const getJwksWithRetries = promiseRetry(
+      async (retry) => {
+        try {
+          const { data } = await axios.get<PublicJwks>(
+            this.#spOidcNdiJwksEndpoint,
+          )
+          return data
+        } catch (e) {
+          return retry(e)
+        }
+      },
+      {
+        retries: 2, // NDI specs: 3 attempts. Do once then retry two times
+        minTimeout: 0, // Retry immediately upon failure
+        maxTimeout: 3000, // NDI specs: timeout of max 3s
+        maxRetryTime: 3000,
+      },
+    )
+
+    const spOidcNdiPublicJwks = await getJwksWithRetries
+
+    return spOidcNdiPublicJwks.keys.map((jwk) => {
+      if (!isEC(jwk) || !jwk.kid || !jwk.use) {
+        throw new JwkError()
+      }
+      return {
+        kid: jwk.kid,
+        use: jwk.use,
+        key: createPublicKey(jwkToPem(jwk)), // Conversion to pem is necessary because in node 14, crypto does not support import of JWK directly
+      }
+    })
+  }
+
+  /**
+   * Method to make network call to NDI's discovery endpoint and construct the base client
+   * Max of 3 attemps with timeout of 3s per attempt
+   * @async
+   * @returns Base client
+   */
+  async retrieveBaseClientFromNdi(): Promise<BaseClient> {
+    const getIssuerWithRetries = promiseRetry(
+      async (retry) => {
+        try {
+          const result = await Issuer.discover(this.#spOidcNdiDiscoveryEndpoint)
+          return result
+        } catch (e) {
+          return retry(e)
+        }
+      },
+      {
+        retries: 2, // NDI specs: 3 attempts. Do once then retry two times
+        minTimeout: 0, // Retry immediately upon failure
+        maxTimeout: 3000, // NDI specs: timeout of max 3s
+        maxRetryTime: 3000,
+      },
+    )
+
+    const issuer = await getIssuerWithRetries
+
+    const baseClient = new issuer.Client(
+      {
+        client_id: this.#spOidcRpClientId,
+        token_endpoint_auth_method: 'private_key_jwt',
+        redirect_uris: [this.#spOidcRpRedirectUrl],
+      },
+      this.#spOidcRpSecretJwks,
+    )
+
+    return baseClient
+  }
+}
+
+/**
+ * Wrapper around the openid-client library to carry out authentication related tasks with Singpass NDI,
+ * and provides methods for decryption and verification of JWE/JWS returned by NDI after authorisation code exchange
+ */
+export class SpOidcClient {
+  #spOidcRpSecretKeys: CryptoKeySet
+  #spOidcRpPublicKeys: CryptoKeySet
```

Comment:
actually no, it was not meant to be a set. I've renamed it to CryptoKeys.

It is not a set because I need to filter on `kid` and `use` values. There may be multiple keys available, but not all of them may be usable.

It is not an object because I have more than one index (`kid` and `use`)